edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/publisher/EdgeRuntimeDiagnosticsUpload.cs

```diff
@@ -25,26 +25,43 @@ public EdgeRuntimeDiagnosticsUpload(IEdgeAgentConnection edgeAgentConnection)
         public async Task<bool> PublishAsync(IEnumerable<Metric> metrics, CancellationToken cancellationToken)
         {
             Preconditions.CheckNotNull(metrics, nameof(metrics));
-            byte[] data = MetricsSerializer.MetricsToBytes(metrics).ToArray();
 
-            if (data.Length > 0)
+            Message[] messagesToSend = this.BatchAndBuildMessages(metrics.ToArray()).ToArray();
+            try
+            {
+                await this.edgeAgentConnection.SendEventBatchAsync(messagesToSend);
+            }
+            catch (Exception ex) when (ex.HasTimeoutException())
             {
-                Message message = this.BuildMessage(data);
-
-                try
-                {
-                    await this.edgeAgentConnection.SendEventAsync(message);
-                }
-                catch (Exception ex) when (ex.HasTimeoutException())
-                {
-                    Log.LogDebug(ex, "Send message to IoTHub");
-                    return false;
-                }
+                Log.LogDebug(ex, "Send message to IoTHub");
+                return false;
             }
 
             return true;
         }
 
+        IEnumerable<Message> BatchAndBuildMessages(ArraySegment<Metric> metrics)
+        {
+            if (!metrics.Any())
+            {
+                return Enumerable.Empty<Message>();
+            }
+
+            byte[] data = MetricsSerializer.MetricsToBytes(metrics).ToArray();
+
+            if (data.Length > 250000)
+            {
+                var part1 = this.BatchAndBuildMessages(metrics.Slice(0, data.Length / 2));
+                var part2 = this.BatchAndBuildMessages(metrics.Slice(data.Length / 2, data.Length / 2));
```

Comment:
part2 would lose one segment if data.length is odd.

Reply:
With recursion, this would lose more in between.

Reply:
You are looking at a very outdated version somehow. I think you are looking at the first commit only. There are tests to check both even and odd versions and that is not an issue in the current version.